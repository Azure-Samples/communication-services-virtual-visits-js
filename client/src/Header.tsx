// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { Stack, Text, ThemeContext, createTheme } from '@fluentui/react';
import { Theme } from '@fluentui/theme';
import { WaffleMenu } from './WaffleMenu';
import { headerContainerStyles, headerTextStyles } from './styles/Header.styles';

/**
 * @interface HeaderProps The properties that Header uses.
 * @member parentid The ID of the element that will be used to contain the panel. This is required so that the panel does not cover the entire height of the screen.
 */
interface HeaderProps {
  companyName: string;
  parentid: string;
}

export const Header = (props: HeaderProps): JSX.Element => {
  const { companyName, parentid } = props;

  return (
    <ThemeContext.Consumer>
      {(theme: Theme | undefined) => {
        if (theme === undefined) {
          theme = createTheme();
        }
        return (
          <Stack styles={headerContainerStyles(theme)} verticalAlign="center" horizontal>
            <WaffleMenu parentid={parentid} />
            <Text styles={headerTextStyles(theme)}>{companyName}</Text>
          </Stack>
        );
      }}
    </ThemeContext.Consumer>
  );
};
