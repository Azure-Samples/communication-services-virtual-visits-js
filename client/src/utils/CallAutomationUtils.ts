// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { CommunicationUserIdentifier } from '@azure/communication-common';
import { CallAdapterState, CommonCallAdapter } from '@azure/communication-react';

export type SummarizeResult = {
  recap: string;
  chapters: {
    chapterTitle: string;
    narrative: string;
  }[];
};

export type CallTranscription = {
  text: string;
  confidence: number;
  participant: CommunicationUserIdentifier;
  resultState: 'intermediate' | 'final';
}[];

/**
 * Fetches the transcription for the given server call ID.
 * @param serverCallId - The server call ID to fetch the transcription for.
 * @returns - the CallTranscription object containing the transcription data.
 */
export const fetchTranscript = async (serverCallId: string): Promise<CallTranscription> => {
  const response = await fetch(`/api/fetchTranscript`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      serverCallId
    })
  });
  if (!response.ok) {
    throw new Error('Failed to fetch transcript: ' + response.statusText);
  }

  return ((await response.json()) as { transcript: CallTranscription }).transcript;
};

/**
 * fetches the transcription status for the given server call ID. This is called when
 * the user first joins the call to check if transcription is already started.
 * @param serverCallId - The server call ID to fetch the transcription status for.
 */
export const fetchTranscriptionStatus = async (serverCallId: string): Promise<void> => {
  const response = await fetch(`/api/fetchTranscriptionState`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      serverCallId
    })
  });
  if (!response.ok) {
    throw new Error('Failed to fetch transcription status: ' + response.statusText);
    return;
  }
  return;
};

/**
 * Starts the transcription for the call.
 * @param serverCallId - The server call ID to start transcription for.
 * @param transcriptionOptions - The options for transcription, including locale.
 * @returns - True if the transcription was started successfully, false otherwise.
 */
export const startTranscription = async (
  serverCallId: string,
  transcriptionOptions?: {
    locale?: string;
  }
): Promise<boolean> => {
  console.log('Starting transcription for call:', serverCallId);
  const response = await fetch('/api/startTranscription', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      serverCallId,
      options: transcriptionOptions
    })
  });
  if (!response.ok) {
    console.error('Failed to start transcription:', response);
    return false;
  }
  console.log('Started transcription:', transcriptionOptions);
  return true;
};

/**
 * Stops the transcription for the call.
 * @param serverCallId - The server call ID to stop transcription for.
 * @returns {boolean} - True if the transcription was stopped successfully, false otherwise.
 */
export const stopTranscription = async (serverCallId: string): Promise<boolean> => {
  console.log('Stopping transcription for call:', serverCallId);

  const response = await fetch('/api/stopTranscription', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      serverCallId
    })
  });
  if (!response.ok) {
    console.error('Failed to stop transcription:', response);
    return false;
  }
  console.log('Stopped transcription');
  return true;
};

/**
 * Connect the Call Automation service to the call.
 * @param callAdaterState - The call adapter state to check for the call.
 * @returns
 */
export const connectToCallAutomation = async (callAdaterState: CallAdapterState): Promise<boolean> => {
  if (callAdaterState.call?.info !== undefined && callAdaterState.call?.state === 'Connected') {
    const serverCallID = await callAdaterState.call.info.getServerCallId();
    const response = await fetch('/api/connectRoomsCall', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        serverCallId: serverCallID
      })
    });
    if (!response.ok) {
      throw new Error('Failed to start call with transcription');
    }
    return true;
  }
  return false;
};

/**
 * Fetch the call summary from the server.
 * @param adapter - The call adapter to get the call ID from.
 * @param locale - locale code for the language to be used in the summary.
 * @returns summary result if there is a summary available.
 */
export const getCallSummaryFromServer = async (
  adapter: CommonCallAdapter,
  locale: LocaleCode
): Promise<SummarizeResult> => {
  console.log('Getting summary from server...');
  try {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const serverCallId = await adapter.getState().call?.info?.getServerCallId();
    if (!serverCallId) {
      console.error('Call ID not found');
      throw new Error('Call ID not found');
    }

    const response = await fetch('/api/summarizeTranscript', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ serverCallId: serverCallId, locale: locale })
    });

    if (!response.ok) {
      console.error('Response Failed: ', response.statusText);
      throw new Error('Summarization request failed');
    }

    const result = await response.json();
    console.log('Summary result:', result);
    return result as SummarizeResult;
  } catch (error) {
    console.error('Error fetching summary:', error);
    throw error;
  }
};

/**
 * Update the participants in the call record for transcription.
 * @param participant - The participant to add to the call record.
 */
export const sendParticipantInfoToServer = async (
  participant: { userId: string; displayName: string },
  serverCallId?: string
): Promise<void> => {
  if (!serverCallId) {
    console.error('Server call ID not found');
    return;
  }
  const response = await fetch('/api/updateParticipants', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      serverCallId: serverCallId,
      participant: participant
    })
  });
  if (!response.ok) {
    console.error('Failed to update participants:', response);
  }
};

/**
 * Fetches the participants in the call. This is needed for when we are computing the transcript for download.
 * @param serverCallId - The server call ID to fetch the participants for.
 * @returns display name and user ID of the participants in the call.
 */
export const fetchParticipants = async (serverCallId: string): Promise<{ userId: string; displayName: string }[]> => {
  const response = await fetch('/api/fetchParticipants', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      serverCallId: serverCallId
    })
  });
  if (!response.ok) {
    console.error('Failed to fetch participants:', response);
    return [];
  }
  return ((await response.json()) as { participants: { userId: string; displayName: string }[] }).participants;
};

/**
 * Language codes used in the localization of the components.
 */
export type LocaleCode =
  | 'ar-AE'
  | 'ar-SA'
  | 'da-DK'
  | 'de-DE'
  | 'en-AU'
  | 'en-CA'
  | 'en-GB'
  | 'en-IN'
  | 'en-NZ'
  | 'en-US'
  | 'es-ES'
  | 'es-MX'
  | 'fi-FI'
  | 'fr-CA'
  | 'fr-FR'
  | 'hi-IN'
  | 'it-IT'
  | 'ja-JP'
  | 'ko-KR'
  | 'nb-NO'
  | 'nl-BE'
  | 'nl-NL'
  | 'pl-PL'
  | 'pt-BR'
  | 'ru-RU'
  | 'sv-SE'
  | 'zh-CN'
  | 'zh-HK'
  | 'cs-CZ'
  | 'pt-PT'
  | 'tr-TR'
  | 'vi-VN'
  | 'th-TH'
  | 'he-IL'
  | 'cy-GB'
  | 'uk-UA'
  | 'el-GR'
  | 'hu-HU'
  | 'ro-RO'
  | 'sk-SK'
  | 'zh-TW';

/**
 * Mapping of locale codes to their display names.
 */
export const localeDisplayNames: Record<LocaleCode, string> = {
  'ar-AE': 'Arabic - U.A.E.',
  'ar-SA': 'Arabic - Saudi Arabia',
  'da-DK': 'Danish',
  'de-DE': 'German - Germany',
  'en-AU': 'English - Australia',
  'en-CA': 'English - Canada',
  'en-GB': 'English - United Kingdom',
  'en-IN': 'English - India',
  'en-NZ': 'English - New Zealand',
  'en-US': 'English - United States',
  'es-ES': 'Spanish - Spain (Modern Sort)',
  'es-MX': 'Spanish - Mexico',
  'fi-FI': 'Finnish',
  'fr-CA': 'French - Canada',
  'fr-FR': 'French - France',
  'hi-IN': 'Hindi',
  'it-IT': 'Italian - Italy',
  'ja-JP': 'Japanese',
  'ko-KR': 'Korean',
  'nb-NO': 'Norwegian (Bokm√•l)',
  'nl-BE': 'Dutch - Belgium',
  'nl-NL': 'Dutch - Netherlands',
  'pl-PL': 'Polish',
  'pt-BR': 'Portuguese - Brazil',
  'ru-RU': 'Russian',
  'sv-SE': 'Swedish',
  'zh-CN': "Chinese - People's Republic of China",
  'zh-HK': 'Chinese - Hong Kong SAR',
  'cs-CZ': 'Czech',
  'pt-PT': 'Portuguese - Portugal',
  'tr-TR': 'Turkish',
  'vi-VN': 'Vietnamese',
  'th-TH': 'Thai',
  'he-IL': 'Hebrew',
  'cy-GB': 'Welsh',
  'uk-UA': 'Ukrainian',
  'el-GR': 'Greek',
  'hu-HU': 'Hungarian',
  'ro-RO': 'Romanian',
  'sk-SK': 'Slovak',
  'zh-TW': 'Chinese - Taiwan'
};
