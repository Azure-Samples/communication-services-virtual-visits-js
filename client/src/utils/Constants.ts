// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

export const BOOKINGS_README_URL = 'https://aka.ms/virtual-appointments-sample-bookings';
export const BOOKINGS_SPECIMEN_URL =
  'https://microsoftbookings.azurewebsites.net/?organization=financialservices&UICulture=en-US';
