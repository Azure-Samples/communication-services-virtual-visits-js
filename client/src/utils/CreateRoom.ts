// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
import { RoomParticipant, RoomParticipantRole } from '../models/RoomModel';
import { createRoom } from './FetchRoomsResponse';
import { makeRoomsJoinUrl } from './GetMeetingLink';

export const createRoomAndRedirectUrl = async (): Promise<string> => {
  let redirectUrl: string;
  try {
    const roomResponse = await createRoom();
    const roomId = roomResponse.roomId;
    const userId = roomResponse.participants.find(
      (participant: RoomParticipant) => (participant.role as RoomParticipantRole) === RoomParticipantRole.presenter
    )?.id;
    if (userId) redirectUrl = makeRoomsJoinUrl(roomId, userId);
    else throw new Error(`room does not have participant with role ${RoomParticipantRole.presenter}`);
  } catch (e) {
    //EndUser to add logging errors in place of console logging below
    console.log(e);
    throw e;
  }
  return redirectUrl;
};
