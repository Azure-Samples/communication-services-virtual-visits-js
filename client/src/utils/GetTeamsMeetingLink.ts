// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { TeamsMeetingLinkLocator } from '@azure/communication-calling';

const MEETING_URL_PARAMNAME = 'meetingURL';
const JOIN_WEB_URL_PARAMNAME = 'JoinWebUrl';
const VISIT_HOSTNAME = 'visit.teams.microsoft.com';
const WEB_JOIN_PARAMNAME = 'webjoin';
const UNIFIED_PARAMNAME = 'unified';

export const getTeamsMeetingLink = (queryString: string): TeamsMeetingLinkLocator => {
  let meetingUrl = new URLSearchParams(queryString).get(MEETING_URL_PARAMNAME);
  if (!meetingUrl) throw 'Unable to get meetingURL from the url string';

  // Urls we get from Bookings are in the form https://visit.teams.microsoft.com?JoinWebUrl=<MEETING_LINK>.
  // The MEETING_LINK also has additional params appended to it which are not essential for joining the meeting. We have
  // to remove these, the unified and webjoin params, because including it causes the MeetingComposite to throw an error
  // about invalid link.
  if (meetingUrl.includes(VISIT_HOSTNAME)) {
    meetingUrl = new URLSearchParams(meetingUrl).get(JOIN_WEB_URL_PARAMNAME);
    if (!meetingUrl) throw 'Unable to get meetingURL from the alternative style url string';

    const url = new URL(meetingUrl);
    url.searchParams.delete(UNIFIED_PARAMNAME);
    url.searchParams.delete(WEB_JOIN_PARAMNAME);
    meetingUrl = url.toString();
  }

  return { meetingLink: meetingUrl };
};

export const getCurrentMeetingURL = (queryString: string): string => {
  let meetingUrl = new URLSearchParams(queryString).get(MEETING_URL_PARAMNAME);
  if (!meetingUrl) {
    meetingUrl = '';
  }
  return meetingUrl;
};

export const getChatThreadIdFromTeamsLink = (teamsMeetingLink: string): string => {
  // Get the threadId from the url - this also contains the call locator ID that will be removed in the threadId.split
  let threadId = teamsMeetingLink.replace('https://teams.microsoft.com/l/meetup-join/', '');
  // Decode characters that outlook links encode
  threadId = threadId.replace(/%3a/g, ':').replace(/%40/g, '@');
  // Extract just the chat guid from the link, stripping away the call locator ID
  threadId = threadId.split(/^(.*?@thread\.(?:v2|skype))/gm)[1];

  if (!threadId || threadId.length === 0) {
    throw new Error('Could not get chat thread from teams link');
  }

  return threadId;
};

export const isValidTeamsLink = (teamsMeetingLink: string): boolean => {
  //If teams link does not have a specific pattern at start, it is invalid
  if (!teamsMeetingLink.startsWith('https://teams.microsoft.com/l/meetup-join/')) {
    return false;
  }
  return true;
};

export const makeJoinUrl = (teamsMeetingLink: string): string =>
  '?' + MEETING_URL_PARAMNAME + '=' + encodeURIComponent(teamsMeetingLink);
