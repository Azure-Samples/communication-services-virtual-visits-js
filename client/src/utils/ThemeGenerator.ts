// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { ThemeGenerator, themeRulesStandardCreator, createTheme, BaseSlots } from '@fluentui/react';
import { Theme } from '@fluentui/theme';

export function generateTheme(primaryColor: string): Theme {
  const themeRules = themeRulesStandardCreator();
  ThemeGenerator.setSlot(themeRules[BaseSlots[BaseSlots.primaryColor]], primaryColor, false, true, true);
  ThemeGenerator.insureSlots(themeRules, false);
  const themeAsJson = ThemeGenerator.getThemeAsJson(themeRules);
  return createTheme({ ...{ palette: themeAsJson } });
}
