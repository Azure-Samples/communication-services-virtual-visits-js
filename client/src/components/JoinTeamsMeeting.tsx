// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import React from 'react';
import { Stack, TextField, PrimaryButton, LayerHost, Theme, ThemeContext, createTheme } from '@fluentui/react';
import { Header } from '../Header';
import { AppConfigModel } from '../models/ConfigModel';
import { getCurrentMeetingURL, isValidTeamsLink, makeJoinUrl } from '../utils/GetTeamsMeetingLink';
import { backgroundStyles } from '../styles/Common.styles';
import {
  makeJoinTeamsLayerHostStyles,
  mainJoinTeamsMeetingContainerStyles,
  mainJoinTeamsMeetingContainerMobileStyles,
  formStyles
} from '../styles/JoinTeamsMeeting.Styles';
import MobileDetect from 'mobile-detect';

interface JoinTeamsMeetingProps {
  config: AppConfigModel;
  onJoinMeeting(urlToJoin: string): void;
}

interface JoinTeamsMeetingState {
  teamsMeetingLink: string;
}

export class JoinTeamsMeeting extends React.Component<JoinTeamsMeetingProps, JoinTeamsMeetingState> {
  public constructor(props: JoinTeamsMeetingProps) {
    super(props);

    this.state = {
      teamsMeetingLink: getCurrentMeetingURL(window.location.search)
    };
  }

  private onTeamsMeetingLinkChange(
    _event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>,
    newValue?: string
  ): void {
    if (newValue) {
      this.setState({ teamsMeetingLink: newValue });
    } else {
      this.setState({ teamsMeetingLink: '' });
    }
  }

  private onGetErrorMessage(value: string): string {
    if (isValidTeamsLink(value) || value === '') {
      return '';
    }

    return 'This meeting link is invalid. Verify your meeting link URL.';
  }

  render(): JSX.Element {
    const link = makeJoinUrl(this.state.teamsMeetingLink);
    const enableButton = isValidTeamsLink(this.state.teamsMeetingLink);
    const parentID = 'JoinTeamsMeetingSection';

    return (
      <ThemeContext.Consumer>
        {(theme: Theme | undefined) => {
          if (theme === undefined) {
            theme = createTheme();
          }

          const containerStyles = new MobileDetect(window.navigator.userAgent).mobile()
            ? mainJoinTeamsMeetingContainerMobileStyles(theme)
            : mainJoinTeamsMeetingContainerStyles(theme);

          return (
            <Stack styles={backgroundStyles(theme)}>
              <Header companyName={this.props.config.companyName} parentid={parentID} />
              <LayerHost id={parentID} style={makeJoinTeamsLayerHostStyles()}>
                <Stack styles={containerStyles} tokens={{ childrenGap: 15 }}>
                  <TextField
                    label="Join a call"
                    placeholder="Enter a meeting link"
                    styles={formStyles}
                    iconProps={{ iconName: 'Link' }}
                    onChange={this.onTeamsMeetingLinkChange.bind(this)}
                    onGetErrorMessage={this.onGetErrorMessage.bind(this)}
                    defaultValue={this.state.teamsMeetingLink}
                  />
                  <PrimaryButton
                    disabled={!enableButton}
                    styles={formStyles}
                    text={'Join call'}
                    onClick={() => this.props.onJoinMeeting(link)}
                  />
                </Stack>
              </LayerHost>
            </Stack>
          );
        }}
      </ThemeContext.Consumer>
    );
  }
}
