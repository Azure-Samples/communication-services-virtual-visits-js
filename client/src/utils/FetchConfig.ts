// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { AppConfigModel } from '../models/ConfigModel';
import { generateTheme } from './ThemeGenerator';
import createError from 'http-errors';

export async function fetchConfig(): Promise<AppConfigModel | undefined> {
  const response = await fetch('/api/config', { method: 'GET' });
  if (response.status !== 200) {
    throw new createError(
      response.status,
      `Unable to retrieve config. Status: ${response.status}. Message: ${response.statusText}.`
    );
  }

  const responseContent = await response.text();
  const config = JSON.parse(responseContent);
  if (!config) {
    throw new Error(`Received empty config response.`);
  }

  const appConfigModel: AppConfigModel = {
    ...config,
    theme: generateTheme(config.colorPalette)
  };

  return appConfigModel;
}
