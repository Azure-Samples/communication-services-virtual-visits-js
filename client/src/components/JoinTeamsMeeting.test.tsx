// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { LayerHost, PrimaryButton, setIconOptions, Stack, TextField } from '@fluentui/react';
import { mount } from 'enzyme';
import { generateTheme } from '../utils/ThemeGenerator';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { JoinTeamsMeeting } from './JoinTeamsMeeting';
import { Header } from '../Header';
import { getTeamsMeetingLink } from '../utils/GetTeamsMeetingLink';
import {
  mainJoinTeamsMeetingContainerMobileStyles,
  mainJoinTeamsMeetingContainerStyles
} from '../styles/JoinTeamsMeeting.Styles';

configure({ adapter: new Adapter() });

// Disable icon warnings for tests as we don't register the icons for unit tests which causes warnings.
// See: https://github.com/microsoft/fluentui/wiki/Using-icons#test-scenarios
setIconOptions({
  disableWarnings: true
});

const validTeamsMeetingLink = getTeamsMeetingLink(
  '?meetingURL=https%3A%2F%2Fteams.microsoft.com%2Fl%2Fmeetup-join%2F19%253ameeting_AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA%2540thread.v2%2F0%3Fcontext%3D%257b%2522Tid%2522%253a%252200000000-0000-0000-0000-000000000000%2522%252c%2522Oid%2522%253a%252200000000-0000-0000-0000-000000000000%2522%257d'
).meetingUrl;

let userAgentGetter: any = undefined;

beforeEach(() => {
  userAgentGetter = jest.spyOn(window.navigator, 'userAgent', 'get');
});

describe('JoinTeamsMeeting', () => {
  it('should render header when page is loaded', async () => {
    const meeting = mount(
      <JoinTeamsMeeting
        config={{
          communicationEndpoint: 'enpoint=test_endpoint;',
          microsoftBookingsUrl: '',
          chatEnabled: true,
          screenShareEnabled: true,
          companyName: '',
          theme: generateTheme('#FFFFFF'),
          waitingTitle: '',
          waitingSubtitle: '',
          logoUrl: ''
        }}
        onJoinMeeting={jest.fn()}
      />
    );

    const headers = meeting.find(Header);

    expect(headers.length).toBe(1);
  });

  it('join button should be disabled when meeting link does not exist', async () => {
    const meeting = mount(
      <JoinTeamsMeeting
        config={{
          communicationEndpoint: 'enpoint=test_endpoint;',
          microsoftBookingsUrl: '',
          chatEnabled: true,
          screenShareEnabled: true,
          companyName: '',
          theme: generateTheme('#FFFFFF'),
          waitingTitle: '',
          waitingSubtitle: '',
          logoUrl: ''
        }}
        onJoinMeeting={jest.fn()}
      />
    );

    const joinButton = meeting.find(PrimaryButton);
    const buttonState = joinButton.prop('disabled');

    expect(buttonState).toBe(true);
  });

  it('should enable join button when meeting link is added', async () => {
    const meeting = mount(
      <JoinTeamsMeeting
        config={{
          communicationEndpoint: 'enpoint=test_endpoint;',
          microsoftBookingsUrl: '',
          chatEnabled: true,
          screenShareEnabled: true,
          companyName: '',
          theme: generateTheme('#FFFFFF'),
          waitingTitle: '',
          waitingSubtitle: '',
          logoUrl: ''
        }}
        onJoinMeeting={jest.fn()}
      />
    );

    meeting.setState({ teamsMeetingLink: validTeamsMeetingLink });
    const joinButton = meeting.find(PrimaryButton);
    const buttonState = joinButton.prop('disabled');

    expect(buttonState).toBe(false);
  });

  it.each([['desktop'], ['mobile']])('uses correct styles if form factor is %s', async (formFactor: string) => {
    const theme = generateTheme('#FFFFFF');
    let expectedStyles = mainJoinTeamsMeetingContainerStyles(theme);

    if (formFactor === 'mobile') {
      const mobileSafariUserAgent =
        'Mozilla/5.0 (iPhone; CPU iPhone OS 14_4 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.3 Mobile/15E148 Safari/604.1';
      userAgentGetter.mockReturnValue(mobileSafariUserAgent);
      expectedStyles = mainJoinTeamsMeetingContainerMobileStyles(theme);
    }

    const meeting = mount(
      <JoinTeamsMeeting
        config={{
          communicationEndpoint: 'enpoint=test_endpoint;',
          microsoftBookingsUrl: '',
          chatEnabled: true,
          screenShareEnabled: true,
          companyName: 'Lamna Healthcare',
          theme: theme,
          waitingTitle: 'Hello',
          waitingSubtitle: 'World',
          logoUrl: ''
        }}
        onJoinMeeting={jest.fn()}
      />
    );

    const container = meeting.find(LayerHost).find(Stack);
    const styles = container.prop('styles');

    expect(styles).toEqual(expectedStyles);
  });

  it('should call onJoinMeeting prop when join button is clicked', async () => {
    const meeting = mount(
      <JoinTeamsMeeting
        config={{
          communicationEndpoint: 'enpoint=test_endpoint;',
          microsoftBookingsUrl: '',
          chatEnabled: true,
          screenShareEnabled: true,
          companyName: '',
          theme: generateTheme('#FFFFFF'),
          waitingTitle: '',
          waitingSubtitle: '',
          logoUrl: ''
        }}
        onJoinMeeting={jest.fn()}
      />
    );

    meeting.setState({ teamsMeetingLink: validTeamsMeetingLink });
    const joinButton = meeting.find(PrimaryButton);

    joinButton.simulate('click');

    expect(meeting.props().onJoinMeeting).toBeCalled();
  });
});

describe('Error handling', () => {
  // FluentUI's TextField doesn't immediatelly call onGetErrorMessage when the value changes.
  // Instead it is called on a timer.
  // So we need fake timers to wait in the test until the callback does get called.
  // https://github.com/microsoft/fluentui/discussions/16240#discussioncomment-217231
  beforeEach(() => {
    // Use modern timers, otherwise jest.runAllTimers() later in the test might fail due to the
    // Lodash's debounce usage in TextField
    // https://github.com/facebook/jest/issues/3465#issuecomment-623393230
    jest.useFakeTimers('modern');
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  const errorMessage = 'This meeting link is invalid. Verify your meeting link URL.';

  it('must show error message when meeting link is invalid', async () => {
    const meeting = mount(
      <JoinTeamsMeeting
        config={{
          communicationEndpoint: 'enpoint=test_endpoint;',
          microsoftBookingsUrl: '',
          chatEnabled: true,
          screenShareEnabled: true,
          companyName: '',
          theme: generateTheme('#FFFFFF'),
          waitingTitle: '',
          waitingSubtitle: '',
          logoUrl: ''
        }}
        onJoinMeeting={jest.fn()}
      />
    );

    const textField = meeting.find(TextField).find('input');
    textField.simulate('change', { target: { value: 'bad meeting link' } });
    jest.runAllTimers();
    const allText = meeting.text();
    expect(allText).toContain(errorMessage);
  });

  it('must not show error message when meeting link is valid', async () => {
    const meeting = mount(
      <JoinTeamsMeeting
        config={{
          communicationEndpoint: 'enpoint=test_endpoint;',
          microsoftBookingsUrl: '',
          chatEnabled: true,
          screenShareEnabled: true,
          companyName: '',
          theme: generateTheme('#FFFFFF'),
          waitingTitle: '',
          waitingSubtitle: '',
          logoUrl: ''
        }}
        onJoinMeeting={jest.fn()}
      />
    );

    const textField = meeting.find(TextField).find('input');
    textField.simulate('change', { target: { value: validTeamsMeetingLink } });
    jest.runAllTimers();
    const allText = meeting.text();
    expect(allText).not.toContain(errorMessage);
  });
});
