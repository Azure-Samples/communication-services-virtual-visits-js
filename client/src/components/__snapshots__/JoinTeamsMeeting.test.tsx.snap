// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JoinTeamsMeeting uses correct styles if form factor is desktop 1`] = `
<div
  className="ms-Stack css-109"
>
  <div
    className="ms-Stack css-110"
  >
    <div>
      <button
        aria-label="Menu Button"
        className="ms-Button ms-Button--icon root-111"
        data-is-focusable={true}
        id="waffle-menu"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        type="button"
      >
        <span
          className="ms-Button-flexContainer flexContainer-112"
          data-automationid="splitbuttonprimary"
        >
          <i
            aria-hidden={true}
            className="ms-Icon root-105 ms-Button-icon icon-114"
            data-icon-name="Waffle"
            style={
              Object {
                "fontFamily": undefined,
              }
            }
          />
        </span>
      </button>
    </div>
    <span
      className="css-119"
    >
      Lamna Healthcare
    </span>
  </div>
  <div
    className="ms-LayerHost"
    id="JoinTeamsMeetingSection"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": "100%",
        "justifyContent": "center",
        "position": "relative",
      }
    }
  >
    <div
      className="ms-Stack css-120"
    >
      <div
        className="ms-TextField root-122"
      >
        <div
          className="ms-TextField-wrapper"
        >
          <label
            className="ms-Label root-133"
            htmlFor="TextField44"
            id="TextFieldLabel46"
          >
            Join a call
          </label>
          <div
            className="ms-TextField-fieldGroup fieldGroup-123"
          >
            <input
              aria-invalid={false}
              aria-labelledby="TextFieldLabel46"
              className="ms-TextField-field field-124"
              id="TextField44"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onInput={[Function]}
              placeholder="Enter a meeting link"
              type="text"
              value=""
            />
            <i
              aria-hidden={true}
              className="icon-134"
              data-icon-name="Link"
            />
          </div>
        </div>
      </div>
      <button
        aria-disabled={true}
        className="ms-Button ms-Button--primary is-disabled root-135"
        data-is-focusable={false}
        disabled={true}
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        type="button"
      >
        <span
          className="ms-Button-flexContainer flexContainer-112"
          data-automationid="splitbuttonprimary"
        >
          <span
            className="ms-Button-textContainer textContainer-113"
          >
            <span
              className="ms-Button-label label-137"
              id="id__49"
            >
              Join call
            </span>
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`JoinTeamsMeeting uses correct styles if form factor is mobile 1`] = `
<div
  className="ms-Stack css-109"
>
  <div
    className="ms-Stack css-110"
  >
    <div>
      <button
        aria-label="Menu Button"
        className="ms-Button ms-Button--icon root-111"
        data-is-focusable={true}
        id="waffle-menu"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        type="button"
      >
        <span
          className="ms-Button-flexContainer flexContainer-112"
          data-automationid="splitbuttonprimary"
        >
          <i
            aria-hidden={true}
            className="ms-Icon root-105 ms-Button-icon icon-114"
            data-icon-name="Waffle"
            style={
              Object {
                "fontFamily": undefined,
              }
            }
          />
        </span>
      </button>
    </div>
    <span
      className="css-119"
    >
      Lamna Healthcare
    </span>
  </div>
  <div
    className="ms-LayerHost"
    id="JoinTeamsMeetingSection"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": "100%",
        "justifyContent": "center",
        "position": "relative",
      }
    }
  >
    <div
      className="ms-Stack css-141"
    >
      <div
        className="ms-TextField root-122"
      >
        <div
          className="ms-TextField-wrapper"
        >
          <label
            className="ms-Label root-133"
            htmlFor="TextField57"
            id="TextFieldLabel59"
          >
            Join a call
          </label>
          <div
            className="ms-TextField-fieldGroup fieldGroup-123"
          >
            <input
              aria-invalid={false}
              aria-labelledby="TextFieldLabel59"
              className="ms-TextField-field field-124"
              id="TextField57"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onInput={[Function]}
              placeholder="Enter a meeting link"
              type="text"
              value=""
            />
            <i
              aria-hidden={true}
              className="icon-134"
              data-icon-name="Link"
            />
          </div>
        </div>
      </div>
      <button
        aria-disabled={true}
        className="ms-Button ms-Button--primary is-disabled root-135"
        data-is-focusable={false}
        disabled={true}
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        type="button"
      >
        <span
          className="ms-Button-flexContainer flexContainer-112"
          data-automationid="splitbuttonprimary"
        >
          <span
            className="ms-Button-textContainer textContainer-113"
          >
            <span
              className="ms-Button-label label-137"
              id="id__62"
            >
              Join call
            </span>
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;
